train/compute/python/tools/trace_link.py

```diff
@@ -0,0 +1,309 @@
+
+        # print("node_edits: ", node_edits)
+
+        for kineto_id, et_id in node_edits:
+            if kineto_id is not None and et_id is not None:
+                et_enhanced[et_id] = kineto_nodes_mapping[kineto_id].end - kineto_nodes_mapping[kineto_id].start
+
+    return et_enhanced
+
+
+if __name__ == "__main__":
+    import argparse
+    
+    parser = argparse.ArgumentParser(description="Link kineto trace with execution trace")
+    parser.add_argument("--et_file", type=str, required=True, help="Path to the execution trace file")
+    parser.add_argument("--kineto_file", type=str, required=True, help="Path to the kineto trace file")
+    parser.add_argument("--annotation", default='DataLoader', type=str, help="User annotation of the iteration step of execution trace")
+    parser.add_argument("--exact_match", default=False, action='store_true', help="Whether to match the traces exactly")
+
+    args = parser.parse_args()
+
+    # Analysis of execution trace
+    with open(args.et_file, 'r') as f:
+        et = ExecutionGraph(json.load(f))
+
+    nodes = et.get_nodes(clean=True)
+
+    # Root node of execution trace is 1-based
+    et_nodes = collect_nodes(nodes[1])
+
+    print('Number of captured ops in execution trace: ', len(et_nodes))
+
+    # Analysis of kineto trace
+    kineto_trace_events = read_dictionary_from_json(args.kineto_file)['traceEvents']
+
+    sorted_kineto_trace_events = sorted(kineto_trace_events, key=lambda kv: kv['ts'])
+
+    kineto_et_events = [event for event in sorted_kineto_trace_events if 'cat' in event and (event['cat'] == 'cpu_op' or event['cat'] == 'user_annotation')]
+
+    kineto_et_segs = []
+    kineto_et_seg = []
+
+    # The choice below normally does not matter for approximate match since we rely on the isomorphism of
+    # the graphs, but for exact match we will use the execution order and then we should be careful
+
+    # Assume that an iteration ends with the specified annotation
+    end_time = -1
+    for event in kineto_et_events:
+        if end_time > 0 and event['ts'] >= end_time:
+            kineto_et_segs.append(kineto_et_seg)
+            kineto_et_seg = []
+            end_time = -1
+
+        if args.annotation in event['name']:
+            kineto_et_seg.append(event)
+            end_time = event['ts'] + event['dur']
+        else:
+            kineto_et_seg.append(event)
+
+    # # Assume that an iteration starts with the specified annotation
+    # for event in kineto_et_events:
+    #     if args.annotation in event['name']:
+    #         kineto_et_segs.append(kineto_et_seg)
+    #         kineto_et_seg = [event]
+    #     else:
+    #         kineto_et_seg.append(event)
+
+    # Find the segment in kineto trace (assuming it contains multiple) with the closest ops to ET 
+    # (usually ET has 3 additional annotation ops for processes/threads)
+    kineto_et_events = find_closest_segment(kineto_et_segs, len(et_nodes) - 3)
```

Comment:
is it guaranteed that ET only has 1 iteration? in case of multiple this won't work right?

Reply:
Yeah, right now we assume only one iteration in ET and we can improve for multiple in the future.